### PR TITLE
Adding index to list of valid routes

### PR DESCRIPTION
**Pull request by RBIII**
Adding index action as it has come up several times with JSON api routing, updating to use `resources` instead of `resource`, and adding a test

### CODE REVIEW

File: test/action_controller_test.rb [hunk 1]
```diff
@@ -226,15 +226,20 @@ def create_order
                          action: "add",
                          id: 1
 
-        assert_generates "/orders/1/relationships/items",
+        assert_generates "/orders/1/relationships/items/1",
```
Comment (vincedevendra) [inline test/action_controller_test.rb]: Ok, I may have been wrong about `resource` being replaceable by `resources`. This is a breaking change. The url needs to remain the way it was.

Comment (vincedevendra) [reply test/action_controller_test.rb]: I think I'd like to just see a show route added, now that I think about it. You're really not getting a collection of things (index), but showing the relationship.

File: test/test_helper.rb [hunk 2]
```diff
@@ -3,7 +3,8 @@
 # Set up the dummy app
 ENV['RAILS_ENV'] = 'test'
 require_relative 'dummy/config/environment'
-ActiveRecord::Migrator.migrate File.expand_path("dummy/db/migrate/", __dir__)
+
+ActiveRecord::MigrationContext.new(File.expand_path("dummy/db/migrate/", __dir__)).migrate
```
Comment (vincedevendra) [inline test/test_helper.rb]: Why was this change necessary?

File: test/action_controller_test.rb (hunk 1 above)
Comment (RBIII) [reply test/action_controller_test.rb]: Cool I'll change it. It's strange to me that the routes generated by `remove` and `replace` wouldn't include the item_id in the url since I'd think those routes would be dealing with a specific item.

File: test/test_helper.rb (hunk 2 above)
Comment (RBIII) [reply test/test_helper.rb]: The `Migrator` got moved into `MigrationContext` in Rails 5.2  (https://github.com/thiagopradi/octopus/issues/477). I didn't see a rails version specified in the gemfile just that it should be between '>= 4.2', '< 6.0'. It defaulted to 5.2.0 and the Orders table wasn't being built locally or on Jenkins with that.

File: test/action_controller_test.rb (hunk 1 above)
Comment (vincedevendra) [reply test/action_controller_test.rb]: So with these relationship routes, you're dealing with the entire relationship, which is expressed as an array of resource identifiers (id and type), never with just one item within it. If that's the sort of the thing you wanted to do, you'd just do a normal resource route.

E.g.,
If a user has many conditions, you might fetch the conditions themselves at `GET /users/{user_id}/conditions`, and the response would look like this:
```
{ 
  "data": [
     {
       "id": "1",
       "type": "conditions",
       "attributes": { "name": "Lupus" }
       "relationships": {...}
     }
  ]
}
```
You might fetch the relationships at `GET /users/{user_id}/relationships/conditions`, and the response would just be 
```
{ 
  "data": [
     {
       "id": "1",
       "type": "conditions",
     }
  ]
}
```
You can `PATCH` to that same route and supply a request body to replace the whole relationship, `POST` to add one or more items from it, or `DELETE` to remove one or more items from it.

File: test/test_helper.rb (hunk 2 above)
Comment (vincedevendra) [reply test/test_helper.rb]: I wonder if we should add a development dependency on Rails 5.1 instead (or keep this change and add a development dependency on Rails 5.2). It doesn't seem right that the tests should just run with whatever version of Rails is sitting around.

Comment (RBIII) [reply test/test_helper.rb]: Fair enough. Any preference?

Comment (vincedevendra) [reply test/test_helper.rb]: Probably 5.1 since that's what plm-website is on.

Comment (RBIII) [reply test/test_helper.rb]: Word updated and left the `MigrationContext` format as a comment in case we ever upgrade rails here

File: Gemfile [hunk 3]
```diff
@@ -1,5 +1,5 @@
 source 'https://rubygems.org'
 
 gemspec
-
+gem 'rails', '5.1.4'
```
Comment (vincedevendra) [inline Gemfile]: Let's specify this in the gemspec intead.

File: lib/simple_jsonapi/rails/extensions/routing.rb [hunk 4]
```diff
@@ -5,6 +5,7 @@ module Routing
         add: :create,
         remove: :destroy,
         replace: :update,
+        show: :show,
```
Comment (mdeutsch) [inline lib/simple_jsonapi/rails/extensions/routing.rb]: The spec doesn't require this, but it refers to this action as "fetching" the relationship. Should we call this "fetch" instead of "show"?

File: simple_jsonapi_rails.gemspec [hunk 5]
```diff
@@ -19,7 +19,7 @@ Gem::Specification.new do |spec|
   spec.require_paths = ['lib']
 
   spec.add_runtime_dependency 'simple_jsonapi'
-  spec.add_runtime_dependency 'rails', '>= 4.2', '< 6.0'
+  spec.add_runtime_dependency 'rails', '5.1.4'
```
Comment (mdeutsch) [inline simple_jsonapi_rails.gemspec]: I don't think we want this restriction as a runtime dependency -- it should only apply to dev/test environments. This may mean that it has to be specified in the Gemfile?

Also, let's make this `'~> 5.1.0'` to allow any 5.1.x version.

Comment (vincedevendra) [inline simple_jsonapi_rails.gemspec]: You'll want to leave this alone, or else you'll force any other folks who use this gem to this version of Rails. But you can add a `spec.add_development_dependency` so that the version of rails that is used locally within the project when running tests is Rails 5.1.4.

File: lib/simple_jsonapi/rails/extensions/routing.rb (hunk 4 above)
Comment (vincedevendra) [reply lib/simple_jsonapi/rails/extensions/routing.rb]: I have no strong opinion about this. It also talks about "fetching" resources, which corresponds to either 'show' or 'index' actions.

Comment (vincedevendra) [reply lib/simple_jsonapi/rails/extensions/routing.rb]: The reason I used the nonstandard action names for the other routes is because they don't map well semantically to any of the standard Rails CRUD actions.

**Comment by RBIII** [comment]
Tried combining answers regarding rails development vs runtime dependency and changed action name for show to `fetch`, though I'm fine with keep it as show as well.
